### ISSUE TITLE
cannot scrape etcd3 https with v0.22.2

### ISSUE THREAD

**Issue by jolson490** [coreos/prometheus-operator#1743]
(I have a K8s cluster on AWS using http for etcd3, and running [v0.19.0 of kube-prometheus](https://github.com/coreos/prometheus-operator/tree/v0.19.0/contrib/kube-prometheus) I'm able to (per the [v19 documentation for monitoring etcd](https://github.com/coreos/prometheus-operator/blob/v0.19.0/contrib/kube-prometheus/docs/Monitoring%20external%20etcd.md)) get the etcd nodes in the cluster to show up on `https://<my-prometheus>/targets`; I customized the `ServiceMonitor ` to specify `scheme: http`.)

But I'm stepping up to [v0.22.2](https://github.com/coreos/prometheus-operator/tree/v0.22.2/contrib/kube-prometheus), in a cluster (named "sandbox-green") where I have etcd3 accepting requests via https, and I've tried various `tlsConfig` (and I read through the comments on #898) but I can't get any etcd targets to show up (nor do I see any instance of "etcd" on `https://<my-prometheus>/config`) - any suggestions would be greatly appreciated, and here's details...

Here's the nodes in my "sandbox-green" cluster:
```
  "master": [
      "192.168.213.199",
      "192.168.213.168",
      "192.168.213.132"
  ],
  "agent": [
      "192.168.213.152",
      "192.168.213.182",
      "192.168.213.178",
      "192.168.213.229",
      "192.168.213.219"
  ],
  "etcd": [
      "192.168.213.212",
      "192.168.213.171",
      "192.168.213.210"
  ]
```

When a master node is started up, here's some of the options I pass to `apiserver`:
```
          - --bind-address=0.0.0.0
          - --etcd-servers=https://127.0.0.1:2379
          - --etcd-cafile=/etc/kubernetes/ssl/ca.pem
          - --etcd-certfile=/etc/kubernetes/ssl/etcd.pem
          - --etcd-keyfile=/etc/kubernetes/ssl/etcd-key.pem
```

The kube-prometheus pods:
```
  core@ip-192-168-213-199 ~ $ kubectl get pods -n monitoring -o wide
  NAME                                   READY     STATUS    RESTARTS   AGE       IP                NODE
  alertmanager-main-0                    2/2       Running   0          1h        10.2.142.145      ip-192-168-213-178.ec2.internal
  grafana-54d8db4bc4-h7zqk               1/1       Running   0          1h        10.2.142.143      ip-192-168-213-178.ec2.internal
  kube-state-metrics-74f8fd79f7-wvn86    4/4       Running   0          1h        10.2.240.203      ip-192-168-213-229.ec2.internal
  node-exporter-7s959                    2/2       Running   0          1h        192.168.213.132   ip-192-168-213-132.ec2.internal
  node-exporter-878l6                    2/2       Running   0          1h        192.168.213.152   ip-192-168-213-152.ec2.internal
  node-exporter-9h4hh                    2/2       Running   0          1h        192.168.213.182   ip-192-168-213-182.ec2.internal
  node-exporter-cbpw7                    2/2       Running   0          1h        192.168.213.178   ip-192-168-213-178.ec2.internal
  node-exporter-l45vz                    2/2       Running   0          1h        192.168.213.168   ip-192-168-213-168.ec2.internal
  node-exporter-qh278                    2/2       Running   0          1h        192.168.213.219   ip-192-168-213-219.ec2.internal
  node-exporter-tffcw                    2/2       Running   0          1h        192.168.213.199   ip-192-168-213-199.ec2.internal
  node-exporter-xk5rf                    2/2       Running   0          1h        192.168.213.229   ip-192-168-213-229.ec2.internal
  prometheus-k8s-0                       3/3       Running   1          34m       10.2.142.149      ip-192-168-213-178.ec2.internal
  prometheus-k8s-1                       3/3       Running   1          34m       10.2.240.204      ip-192-168-213-229.ec2.internal
  prometheus-operator-769f6f97cb-wvlm2   1/1       Running   0          1h        10.2.142.142      ip-192-168-213-178.ec2.internal
```

A snippet of the output from running `openssl x509 -noout -text -in /etc/kubernetes/ssl/etcd.pem` on agent node `192.168.213.178` (i.e. the agent that the `prometheus-k8s-0` pod is running on); etcd3 doesn't allow/support DNS, so there is no server name - thus SANs (Subject Alternative Names) are used:
```
            X509v3 Subject Alternative Name:
                DNS:localhost, IP Address:127.0.0.1, IP Address:192.168.213.171, IP Address:192.168.213.210, IP Address:192.168.213.212
```

Note that PR #1732 proposes some revisions to the commentary in kube-prometheus about monitoring etcd (see [README.md](https://github.com/jolson490/prometheus-operator/blob/0ca23ce/contrib/kube-prometheus/README.md#static-etcd-configuration) & [etcd.jsonnet](https://github.com/jolson490/prometheus-operator/blob/0ca23ce/contrib/kube-prometheus/examples/etcd.jsonnet)).

Here's the relevant objects (note that until the "TODO" comment mentioned [here](https://github.com/jolson490/prometheus-operator/blob/0ca23ce/contrib/kube-prometheus/examples/etcd.jsonnet#L27) is resolved, in my own copy of [vendor/kube-prometheus/kube-prometheus-static-etcd.libsonnet](https://github.com/coreos/prometheus-operator/blob/v0.22.2/contrib/kube-prometheus/jsonnet/kube-prometheus/kube-prometheus-static-etcd.libsonnet) I've customized it to not specify a value for `serverName` but instead specify `insecureSkipVerify: true,`):
```
core@ip-192-168-213-199 ~ $ kubectl -n kube-system describe service etcd
Name:              etcd
Namespace:         kube-system
Labels:            k8s-app=etcd
Annotations:       <none>
Selector:          <none>
Type:              ClusterIP
IP:                None
Port:              metrics  2379/TCP
TargetPort:        2379/TCP
Endpoints:         192.168.213.171:2379,192.168.213.210:2379,192.168.213.212:2379
Session Affinity:  None
Events:            <none>
core@ip-192-168-213-199 ~ $ kubectl -n kube-system describe endpoints etcd
Name:         etcd
Namespace:    kube-system
Labels:       k8s-app=etcd
Annotations:  <none>
Subsets:
  Addresses:          192.168.213.171,192.168.213.210,192.168.213.212
  NotReadyAddresses:  <none>
  Ports:
    Name     Port  Protocol
    ----     ----  --------
    metrics  2379  TCP

Events:  <none>
core@ip-192-168-213-199 ~ $ kubectl -n kube-system describe servicemonitor etcd
Name:         etcd
Namespace:    kube-system
Labels:       k8s-app=etcd
Annotations:  <none>
API Version:  monitoring.coreos.com/v1
Kind:         ServiceMonitor
Metadata:
  Cluster Name:
  Creation Timestamp:  2018-08-03T12:42:08Z
  Generation:          1
  Resource Version:    293799
  Self Link:           /apis/monitoring.coreos.com/v1/namespaces/kube-system/servicemonitors/etcd
  UID:                 a2a0be20-971a-11e8-a8e9-0aaf6efae5ec
Spec:
  Endpoints:
    Interval:  30s
    Port:      metrics
    Scheme:    https
    Tls Config:
      Ca File:               /etc/prometheus/secrets/kube-etcd-client-certs/etcd-client-ca.crt
      Cert File:             /etc/prometheus/secrets/kube-etcd-client-certs/etcd-client.crt
      Insecure Skip Verify:  true
      Key File:              /etc/prometheus/secrets/kube-etcd-client-certs/etcd-client.key
  Job Label:                 k8s-app
  Selector:
    Match Labels:
      K 8 S - App:  etcd
Events:             <none>
core@ip-192-168-213-199 ~ $ kubectl -n monitoring describe secret kube-etcd-client-certs
Name:         kube-etcd-client-certs
Namespace:    monitoring
Labels:       <none>
Annotations:  <none>

Type:  Opaque

Data
====
etcd-client-ca.crt:  1069 bytes
etcd-client.crt:     1252 bytes
etcd-client.key:     1703 bytes
```

In the `*.jsonnet` file that I pass to `build.sh`, for the prometheus spec I specified `logLevel: "debug",`. But in `kubectl -n monitoring logs prometheus-k8s-0` I don't see any errors - and the only thing relevant to etcd is lines such as the following:
`level=debug ts=2018-08-03T12:42:40.098677772Z caller=kubernetes.go:385 component="discovery manager scrape" discovery=k8s role=endpoint msg="kubernetes discovery update" role=endpoints tg="&targetgroup.Group{Targets:[]model.LabelSet{model.LabelSet{\"__address__\":\"192.168.213.171:2379\", \"__meta_kubernetes_endpoint_port_name\":\"metrics\", \"__meta_kubernetes_endpoint_port_protocol\":\"TCP\", \"__meta_kubernetes_endpoint_ready\":\"true\"}, model.LabelSet{\"__address__\":\"192.168.213.210:2379\", \"__meta_kubernetes_endpoint_port_name\":\"metrics\", \"__meta_kubernetes_endpoint_port_protocol\":\"TCP\", \"__meta_kubernetes_endpoint_ready\":\"true\"}, model.LabelSet{\"__address__\":\"192.168.213.212:2379\", \"__meta_kubernetes_endpoint_port_name\":\"metrics\", \"__meta_kubernetes_endpoint_port_protocol\":\"TCP\", \"__meta_kubernetes_endpoint_ready\":\"true\"}}, Labels:model.LabelSet{\"__meta_kubernetes_namespace\":\"kube-system\", \"__meta_kubernetes_endpoints_name\":\"etcd\", \"__meta_kubernetes_service_name\":\"etcd\", \"__meta_kubernetes_service_label_k8s_app\":\"etcd\"}, Source:\"endpoints/kube-system/etcd\"}"`

Is there something I'm doing wrong? Or is etcd monitoring somehow broken in `v0.22.2`?

**Comment by jolson490** [comment]
I added `127.0.0.1` (one of the SANs from my etcd cert) to the `Tls Config:` in the `ServiceMonitor ` named `etcd`:
```
    Tls Config:
      Ca File:               /etc/prometheus/secrets/kube-etcd-client-certs/etcd-client-ca.crt
      Cert File:             /etc/prometheus/secrets/kube-etcd-client-certs/etcd-client.crt
      Insecure Skip Verify:  true
      Key File:              /etc/prometheus/secrets/kube-etcd-client-certs/etcd-client.key
      Server Name:           127.0.0.1
```

And then (for good measure) I did `kubectl delete pod` on the `prometheus-*` pods. But still no etcd targets showing up.

By the way I did notice instances of the following in the prometheus-operator log:
```
level=info ts=2018-08-03T16:40:27.173477576Z caller=operator.go:737 component=prometheusoperator msg="sync prometheus" key=monitoring/k8s
W0803 16:40:27.173600       1 listers.go:63] can not retrieve list of objects using index : Index with name namespace does not exist
W0803 16:40:27.173811       1 listers.go:63] can not retrieve list of objects using index : Index with name namespace does not exist
W0803 16:40:27.377410       1 listers.go:63] can not retrieve list of objects using index : Index with name namespace does not exist
```

**Comment by jolson490** [comment]
@brancz , [this comment on issue 898](https://github.com/coreos/prometheus-operator/issues/898#issuecomment-358300264) talks about running prometheus on a master node - was that a feature that ended up not getting added/needed into prometheus-operator, or is that an option/idea of something I can try setting up myself?

**Comment by jolson490** [comment]
I believe the key to figuring this out is to replicate how prometheus scrapes metrics from etcd nodes.

Just curious: can someone please tell me how exactly does prometheus do that? The prometheus pod doesn't have `curl` installed, and the version of `wget` installed seems to be an old/insufficient one that doesn't have options such as `--ca-certificate` - but that probably doesn't matter to prometheus-operator because (just a guess) in its go code it probably creates an https connection/request?

--------

The following `wget` command worked on an agent node (and it shows I do indeed have prometheus using the correct 3 `*.pem` files):
```
core@ip-192-168-213-199 ~/tmp $ kubectl -n monitoring get secret kube-etcd-client-certs -ojson | jq -r '.data["etcd-client-ca.crt"]' | base64 -d > etcd-client-ca.crt
core@ip-192-168-213-199 ~/tmp $ kubectl -n monitoring get secret kube-etcd-client-certs -ojson | jq -r '.data["etcd-client.crt"]' | base64 -d > etcd-client.crt
core@ip-192-168-213-199 ~/tmp $ kubectl -n monitoring get secret kube-etcd-client-certs -ojson | jq -r '.data["etcd-client.key"]' | base64 -d > etcd-client.key

core@ip-192-168-213-199 ~/tmp $ wget -O /home/core/tmp/metrics.txt --ca-certificate=etcd-client-ca.crt --certificate=etcd-client.crt --private-key=etcd-client.key https://192.168.213.212:2379/metrics
--2018-08-03 18:22:52--  https://192.168.213.212:2379/metrics
Connecting to 192.168.213.212:2379... connected.
HTTP request sent, awaiting response... 200 OK
Length: 132310 (129K) [text/plain]
Saving to: '/home/core/tmp/metrics.txt'

/home/core/tmp/metrics.txt                       100%[=============================================================================================================>] 129.21K  --.-KB/s    in 0.001s

2018-08-03 18:22:52 (205 MB/s) - '/home/core/tmp/metrics.txt' saved [132310/132310]
core@ip-192-168-213-199 ~/tmp $ head -5 metrics.txt
# HELP etcd_debugging_mvcc_db_compaction_keys_total Total number of db keys compacted.
# TYPE etcd_debugging_mvcc_db_compaction_keys_total counter
etcd_debugging_mvcc_db_compaction_keys_total 330782
# HELP etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds Bucketed histogram of db compaction pause duration.
# TYPE etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds histogram
```

The `wget` command works not only for an etcd node IP (`192.168.213.212`) but also for `127.0.0.1`:

```
core@ip-192-168-213-199 ~/tmp $ wget -O /home/core/tmp/metrics_127.txt --ca-certificate=etcd-client-ca.crt --certificate=etcd-client.crt --private-key=etcd-client.key https://127.0.0.1:2379/metrics
...
core@ip-192-168-213-199 ~/tmp $ head -5 metrics_127.txt
# HELP etcd_debugging_mvcc_db_compaction_keys_total Total number of db keys compacted.
# TYPE etcd_debugging_mvcc_db_compaction_keys_total counter
etcd_debugging_mvcc_db_compaction_keys_total 330782
# HELP etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds Bucketed histogram of db compaction pause duration.
# TYPE etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds histogram
```

`curl` works as well (note that when running this command on an agent against `127.0.0.1`, I don't need to specify `--insecure`):
```
core@ip-192-168-213-199 /etc/kubernetes/ssl $ curl --silent --cacert ca.pem --cert etcd.pem --key etcd-key.pem https://127.0.0.1:2379/metrics | head -8
# HELP etcd_debugging_mvcc_db_compaction_keys_total Total number of db keys compacted.
# TYPE etcd_debugging_mvcc_db_compaction_keys_total counter
etcd_debugging_mvcc_db_compaction_keys_total 283603
# HELP etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds Bucketed histogram of db compaction pause duration.
# TYPE etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds histogram
etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds_bucket{le="1"} 0
etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds_bucket{le="2"} 0
etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds_bucket{le="4"} 0
```

--------

A couple services:
```
core@ip-192-168-213-199 ~ $ kubectl get services --all-namespaces | egrep "alertmanager-main|etcd"
kube-system                etcd                                           ClusterIP   None         <none>        2379/TCP                      7h
monitoring                 alertmanager-main                              NodePort    10.3.0.182   <none>        9093:30903/TCP                7h
core@ip-192-168-213-199 ~ $ kubectl -n monitoring describe service alertmanager-main | grep Endpoints
Endpoints:                10.2.142.145:9093
core@ip-192-168-213-199 ~ $ kubectl -n kube-system describe service etcd | grep Endpoints
Endpoints:         192.168.213.171:2379,192.168.213.210:2379,192.168.213.212:2379
```

A co-worker helped me setup a test pod...

DNS for the services:
```
root@test:~# nslookup alertmanager-main.monitoring.svc.cluster.local
Server:		10.3.0.10
Address:	10.3.0.10#53

Name:	alertmanager-main.monitoring.svc.cluster.local
Address: 10.3.0.182

root@test:~# nslookup etcd.kube-system.svc.cluster.local
Server:		10.3.0.10
Address:	10.3.0.10#53

Name:	etcd.kube-system.svc.cluster.local
Address: 192.168.213.210
Name:	etcd.kube-system.svc.cluster.local
Address: 192.168.213.212
Name:	etcd.kube-system.svc.cluster.local
Address: 192.168.213.171
```

It's easy to use the DNS to access the alertmanager metrics:
```
root@test:~# curl --silent http://alertmanager-main.monitoring.svc.cluster.local:9093/metrics -o am.txt
root@test:~# head -5 am.txt
# HELP alertmanager_alerts How many alerts by state.
# TYPE alertmanager_alerts gauge
alertmanager_alerts{state="active"} 18
alertmanager_alerts{state="suppressed"} 0
# HELP alertmanager_alerts_invalid_total The total number of received alerts that were invalid.
```

When using the DNS to access the etcd metrics, it only works once adding `--insecure`:
```
# The following command is very similar to a command above that worked on the agent node `ip-192-168-213-199` (the difference is these DNS names don't exist on the agent nodes).
root@test:~# curl --cacert ca.pem --cert etcd.pem --key etcd-key.pem https://etcd.kube-system.svc.cluster.local:2379/metrics
curl: (51) SSL: no alternative certificate subject name matches target host name 'etcd.kube-system.svc.cluster.local'

root@test:~# curl --silent --insecure --cacert ca.pem --cert etcd.pem --key etcd-key.pem https://etcd.kube-system.svc.cluster.local:2379/metrics -o etcd.txt
root@test:~# head -8 etcd.txt
# HELP etcd_debugging_mvcc_db_compaction_keys_total Total number of db keys compacted.
# TYPE etcd_debugging_mvcc_db_compaction_keys_total counter
etcd_debugging_mvcc_db_compaction_keys_total 338007
# HELP etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds Bucketed histogram of db compaction pause duration.
# TYPE etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds histogram
etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds_bucket{le="1"} 0
etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds_bucket{le="2"} 0
etcd_debugging_mvcc_db_compaction_pause_duration_milliseconds_bucket{le="4"} 0
```

**Comment by brancz** [comment]
Thanks for the very detailed report, this is awesome!

We think there might be a bug in 0.22.2 that might cause this, could you try to verify that this also doesn't work with 0.21.0 and 0.22.0?

Also, could you share your `Prometheus` object? I just want to make sure that the `ServiceMonitor` for etcd is properly being selected (as it's in a different namespace).

Out of curiosity, when you issued the wget or curl commands, was that from the host that runs etcd itself or from your local machine?

**Comment by jolson490** [comment]
You're welcome - and thanks for the info.

I'll try those two versions and let you know...

These are the only customizations I've made to the `Prometheus` object:
```
~/git_workspace/my-prometheus-operator
josh@MyMac$ diff ~/gh_monitoring/coreos/v0.22.2_prometheus-operator/prometheus-operator/contrib/kube-prometheus/manifests/prometheus-prometheus.yaml manifests/
14a15,17
>   externalLabels:
>     cluster: sandbox-green
24a28,29
>   secrets:
>   - kube-etcd-client-certs
```

We have etcd itself running on the etcd nodes. The terminal snippets that show commands that were run on `core@ip-192-168-213-199` were run on one of the master nodes in the K8s cluster, and `root@test` was the test pod we created via doing `kubectl run -it --rm --restart=Never test --image=debian bash` on that master node.

**Comment by jolson490** [comment]
I tried 0.22.0 & 0.21.0... (to see if this issue with etcd targets not showing up also occurs with either/both of those releases) - note that I had to manually workaround the following:
 * #1755 "kube-prometheus: re-introduce insecureSkipVerify for etcd monitoring"
 * #1757 "jb install doesn't get the specified prometheusOperator version"
 * [kube-prometheus-static-etcd.libsonnet](https://github.com/coreos/prometheus-operator/blob/master/contrib/kube-prometheus/jsonnet/kube-prometheus/kube-prometheus-static-etcd.libsonnet) was first [introduced](https://github.com/coreos/prometheus-operator/pull/1552/files) in [v0.22.0](https://github.com/coreos/prometheus-operator/tree/v0.22.0) - so for testing 0.21.0 I had to manually merge the changes made by that PR into my `vendor/` folder.

(The `Tls Config:` is still the same as shown in [this comment](https://github.com/coreos/prometheus-operator/issues/1743#issuecomment-410313124).)

...unfortunately etcd targets didn't show up with either of those two versions. Frederic I'll be looking forward to if you can please advise further at your soonest convenience - here's the only ideas I have at this point:
 * Would you like to see my manifests files I generated for 0.22.0 and/or 0.21.0 (just to make sure they're correct)?
 * Should I try any other older prometheus-operator versions?
 * Perhaps what would be best is if you can reproduce this issue?

By the way, it seems like the logging in prometheus/prometheus-operator for scraping is basically non-existent (even with `logLevel: "debug"` for the prometheus spec). Do you think it would be worthwhile if I opened a separate issue (in either [prometheus-operator](https://github.com/coreos/prometheus-operator) or [prometheus](https://github.com/prometheus/prometheus)?) for the scraping logging to be enhanced / made more verbose? It's been very difficult for me to try to diagnose why prometheus isn't finding the etcd targets.

**Comment by brancz** [comment]
I think this has been confusing for a while, and we should document this in the kube-prometheus readme: kube-prometheus is not versioned at all. Any versioned tag on this repository is only ever about the Prometheus Operator, and never about the kube-prometheus sub-project. In fact we want to extract kube-prometheus into a separate repository as soon as the helm charts have been extracted to the upstream stable charts.

> By the way, it seems like the logging in prometheus/prometheus-operator for scraping is basically non-existent (even with logLevel: "debug" for the prometheus spec). Do you think it would be worthwhile if I opened a separate issue (in either prometheus-operator or prometheus?) for the scraping logging to be enhanced / made more verbose? It's been very difficult for me to try to diagnose why prometheus isn't finding the etcd targets.

We can't do anything about this in the Prometheus Operator and it is intentional that these things are not being logged in Prometheus, simply because scrape failures are somewhat expected, and actual scrape errors are displayed on the `/targets` page.

We are about to publish v0.23.0 that should fix cross namespace problems. That said, I think you are mixing a couple of concepts, that we should solve individually:

First let's make sure etcd is actually discovered, regardless of whether it is successfully scraped, if Prometheus can discover the nodes, then we're half way there, and it's only a certificate/authentication problem. Let's first make sure that you are discovering the nodes correctly. Can you see the etcd nodes themselves in the `/targets` page?

---

As a side note, thanks a lot for all your involvement, it is invaluable to us. I will follow up on the pull requests and issues you linked to separately in those issues/pull requests. Thanks a ton for everything!

**Comment by jolson490** [comment]
Thanks Frederic for the info.

Just curious, when you [said](https://github.com/coreos/prometheus-operator/issues/1743#issuecomment-410426605) "there might be a bug in 0.22.2" - do you believe the bug is in [prometheus-operator](https://github.com/coreos/prometheus-operator)?  Not in [kube-prometheus](https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus) nor [prometheus](https://github.com/prometheus/prometheus)? 
And can you please tell me which file(s) in [manifests](https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus/manifests) you believe the bug ends up in? I just wanna make sure I am indeed correctly testing versions 0.22.0 & 0.21.0.

What you said about etcd targets being discovered vs successfully scraped does make sense - what I meant by I can't "get the etcd nodes in the cluster to show up on `https://<my-prometheus>/targets`" is I haven't even gotten prometheus to discover the etcd nodes.

And you're very welcome - and thanks for the continued assistance; we're in this together. :) It's been hard work for me just to keep up with all the changes - but I really like the direction kube-prometheus has gone in and is going in.

**Comment by brancz** [comment]
Ok that all makes sense, thanks for clarification, ok that does sound like there is a problem discovering, and this might be related to bugs that were present in both of those versions, which we fixed in 0.23.0, let's try 0.23.0 once the release is done and we update it in kube-prometheus. All of this should be happening today. Thanks for your patience! :slightly_smiling_face:

**Comment by jolson490** [comment]
just fyi, for my own personal reassurance I deployed [0.19.0](https://github.com/coreos/prometheus-operator/releases/tag/v0.19.0) and followed the [v19 etcd steps](https://github.com/coreos/prometheus-operator/blob/v0.19.0/contrib/kube-prometheus/docs/Monitoring%20external%20etcd.md) (including doing `insecureSkipVerify: true`) and it worked on the first try. :) (Unlike Lego Batman, I honestly can say first try.)

![screen shot 2018-08-07 at 9 25 25 am](https://user-images.githubusercontent.com/15840507/43782079-164397d4-9a24-11e8-995a-b57f8b1bf343.png)

![batman-first-try-stylized_408](https://user-images.githubusercontent.com/15840507/43782097-1ea03c84-9a24-11e8-97fc-046ef924a126.png)

**Comment by jolson490** [comment]
Is there a way I can sign up to be notified in general whenever a new prometheus-operator release is fully ready?

I see 0.23.0 is showing up on https://github.com/coreos/prometheus-operator/releases, and v0.23.0 is on https://quay.io/repository/coreos/prometheus-operator?tab=tags - so is 0.23.0 fully ready for me to try out?

**Comment by squat** [comment]
@jolson490 the v0.23.0 prometheus-operator container itself is ready, but the kube-prometheus stack needs https://github.com/coreos/prometheus-operator/pull/1762 to land before the jsonnet is ready. It will be merged soon!

**Comment by jolson490** [comment]
hi @brancz & @squat . Now that #1762 is merged, I've run into 3 new issues:

1) Sorry if I'm still confused about `jb install` and releases, but:
 * the `kube-prometheus/jsonnetfile.json` file on the [master](https://github.com/coreos/prometheus-operator/blob/fb99938/contrib/kube-prometheus/jsonnet/kube-prometheus/jsonnetfile.json#L41) branch (per one of the [changes](https://github.com/coreos/prometheus-operator/pull/1762/files#diff-2d2c70128d54ca18fef742fdb131b4de) made by PR #1762 ) specifies  "`v0.23.0`" for `prometheus-operator` (which seems good), 
 * but should that file on the [v0.23.0 branch](https://github.com/coreos/prometheus-operator/tree/v0.23.0) also specify "`v0.23.0`" (currently it [specifies v0.22.2](https://github.com/coreos/prometheus-operator/blob/v0.23.0/contrib/kube-prometheus/jsonnet/kube-prometheus/jsonnetfile.json#L41))? 

As things currently stand, if I do `jb install github.com/coreos/prometheus-operator/contrib/kube-prometheus/jsonnet/kube-prometheus/@v0.23.0` then in my `jsonnetfile.lock.json` for `prometheus-operator` I get the [commit](https://github.com/coreos/prometheus-operator/commit/87000af) from `0.22.2`.
Perhaps this relates back to #1757 / #1754 ?
(I worked around this by taking off the "`/@v0.23.0`" from the end of my "`jb install`" command.)

2) Then when I tried to do `kubectl create -f manifests/` I encountered:
```
...
servicemonitor.monitoring.coreos.com "kubelet" created
The Prometheus "k8s" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"monitoring.coreos.com/v1", "kind":"Prometheus", "metadata":map[string]interface {}{"labels":map[string]interface {}{"prometheus":"k8s"}, "name":"k8s", "namespace":"monitoring", "generation":1, "creationTimestamp":"2018-08-07T17:06:13Z", "uid":"303601e5-9a64-11e8-a8e9-0aaf6efae5ec", "selfLink":"", "clusterName":""}, "spec":map[string]interface {}{"baseImage":"quay.io/prometheus/prometheus", "externalUrl":"https://prometheus-sbx-1.pearsonvue.com", "serviceMonitorNamespaceSelector":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{}}}, "version":"v2.3.1", "alerting":map[string]interface {}{"alertmanagers":[]interface {}{map[string]interface {}{"name":"alertmanager-main", "namespace":"monitoring", "port":"web"}}}, "logLevel":"debug", "externalLabels":map[string]interface {}{"cluster":"sandbox-green"}, "nodeSelector":map[string]interface {}{"beta.kubernetes.io/os":"linux"}, "replicas":2, "ruleSelector":map[string]interface {}{"matchLabels":map[string]interface {}{"prometheus":"k8s", "role":"alert-rules"}}, "secrets":[]interface {}{"kube-etcd-client-certs"}, "serviceAccountName":"prometheus-k8s", "resources":map[string]interface {}{"requests":map[string]interface {}{"memory":"400Mi"}}, "serviceMonitorSelector":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"operator":"Exists", "key":"k8s-app"}}}}}: validation failure list:
spec.serviceMonitorNamespaceSelector.matchExpressions.key in body is required
spec.serviceMonitorNamespaceSelector.matchExpressions.operator in body is required
```
Reference info:
 * I see `serviceMonitorNamespaceSelector` has existed at least since [v0.19.0](https://github.com/coreos/prometheus-operator/blob/v0.19.0/Documentation/api.md#prometheusspec).
 * PR 1762 [added 3 lines](https://github.com/coreos/prometheus-operator/pull/1762/files#diff-9ce1faf7c759768f42198063e90d92ad) to `kube-prometheus/manifests/prometheus-prometheus.yaml` for `serviceMonitorNamespaceSelector`.
 * I found [this comment](https://github.com/coreos/prometheus-operator/pull/1227#issue-182199391) on the PR back in April that added `ServiceMonitorNamespaceSelector`.

(I worked around this by commenting out those 3 lines from my `manifests/prometheus-prometheus.yaml`.)

3) Then I did get the latest contents of kube-prometheus from the master branch up & running, but unfortunately prometheus still doesn't discover the etcd nodes :( So perhaps the changes in 0.23.0 didn't fix this issue #1743 after all?
Would you guys please try to reproduce this #1743 and debug it from there?

**Comment by brancz** [comment]
```
spec.serviceMonitorNamespaceSelector.matchExpressions.key in body is required
spec.serviceMonitorNamespaceSelector.matchExpressions.operator in body is required
```

is being fixed with #1771 

Note that `jb install` installs whatever is locked in your `jsonnetfile.lock.json`, if that file exists. To update the dependencies you can either manually bump the commit in that file or run `jb update`.

**Comment by jolson490** [comment]
hi Frederic, per "3." in my most recent comment above, would you accept this GH issue being assigned to you?

**Comment by brancz** [comment]
Sorry for the radio silence most of the team was out for PromCon last week. We're now working off the backlog that has accumulated :slightly_smiling_face: .

**Comment by jolson490** [comment]
Hmm, maybe this actually has been fixed - I just did the following:
 * grabbed the latest from master for this repo (i.e. [v0.23.1](
https://github.com/coreos/prometheus-operator/tree/v0.23.1)) - and modified my own copy to do `http` for etcd
 * deployed it to a K8s cluster that has etcd setup to accept requests over `http` (not `https`)
 * and etcd targets showed up green on the prom targets page

I'm about to also try with a cluster that has etcd using `https`...

**Comment by brancz** [comment]
#1756 has also now been merged, so all https configurations should now be available, let me know how it goes and how we should continue to get the documentation PR merged! :slightly_smiling_face: Once again thanks a lot for your dedication!

**Comment by jolson490** [comment]
I don't know why it didn't work last week ("3." in [this comment](https://github.com/coreos/prometheus-operator/issues/1743#issuecomment-411145087) shows I was using `prometheus-operator:v0.23.0`), but I just tried it now on a cluster using https for etcd and it worked there too. So Frederic (please let me know if I'm missing anything but) I don't believe there's any remaining loose ends here for this issue, so I'm closing it.
